homeassistant/components/iotawatt/coordinator.py

```diff
@@ -0,0 +1,56 @@
+"""IoTaWatt DataUpdateCoordinator."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from iotawattpy.iotawatt import Iotawatt
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import httpx_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONNECTION_ERRORS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IotawattUpdater(DataUpdateCoordinator):
+    """Class to manage fetching update data from the IoTaWatt Energy Device."""
+
+    api: Iotawatt | None = None
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize IotaWattUpdater object."""
+        self.entry = entry
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=entry.title,
+            update_interval=timedelta(seconds=30),
+        )
+
+    async def _async_update_data(self):
+        """Fetch sensors from IoTaWatt device."""
+        if self.api is None:
+            api = Iotawatt(
+                self.entry.title,
+                self.entry.data[CONF_HOST],
+                httpx_client.get_async_client(self.hass),
+                self.entry.data.get(CONF_USERNAME),
+                self.entry.data.get(CONF_PASSWORD),
+            )
+            try:
+                is_authenticated = await api.connect()
+            except CONNECTION_ERRORS as err:
+                raise UpdateFailed("Connection failed") from err
+
+            if not is_authenticated:
+                raise UpdateFailed("Authentication error")
```

Comment:
A future improvement would be to raise `ConfigEntryAuthFailed` and start a reauth config flow.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication